File: src/main/java/site/gachontable/gachontablebe/global/config/aop/RedissonLockAop.java
package site.gachontable.gachontablebe.global.config.aop;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.redisson.api.RLock;
import org.redisson.api.RedissonClient;
import org.springframework.stereotype.Component;
import site.gachontable.gachontablebe.global.config.parser.CustomSpringELParser;
import site.gachontable.gachontablebe.global.config.redis.RedissonLock;

import java.lang.reflect.Method;

@Aspect
@Component
@RequiredArgsConstructor
@Slf4j
public class RedissonLockAop {
    private static final String REDISSON_LOCK_PREFIX = "LOCK:";

    private final RedissonClient redissonClient;
    private final AopForTransaction aopForTransaction;

    @Around("@annotation(site.gachontable.gachontablebe.global.config.redis.RedissonLock)")
    public Object lock(final ProceedingJoinPoint joinPoint) throws Throwable {
        MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        Method method = signature.getMethod();
        RedissonLock redissonLock = method.getAnnotation(RedissonLock.class);

        String key = REDISSON_LOCK_PREFIX + CustomSpringELParser
                .getDynamicValue(signature.getParameterNames(), joinPoint.getArgs(), redissonLock.key());
        RLock rLock = redissonClient.getLock(key);

        try {
            boolean available = rLock.tryLock(redissonLock.waitTime(), redissonLock.leaseTime(), redissonLock.timeUnit());
            if (!available) {
                log.info("Lock 획득 실패 : {}", key);
                return false;
            }

            return aopForTransaction.proceed(joinPoint, key);
        } catch (InterruptedException e) {
            throw new InterruptedException();
        } finally {
            try {
                log.info("Lock 해제 : {}", key);
                rLock.unlock();
            } catch (IllegalMonitorStateException e) {
                log.info("이미 해제된 Lock : {} {}", method.getName(), key);
            }
        }
    }
}

File: src/main/kotlin/site/gachontable/infra/redis/aop/RedissonLockAop.kt
package site.gachontable.infra.redis.aop

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.redisson.api.RedissonClient
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import site.gachontable.independent.parser.CustomSpringELParser.getDynamicValue
import site.gachontable.infra.redis.RedissonLock

@Aspect
@Component
class RedissonLockAop(
    private val redissonClient: RedissonClient,
    private val aopForTransaction: AopForTransaction,
) {
    @Around("@annotation(site.gachontable.infra.redis.RedissonLock)")
    @Throws(Throwable::class)
    fun lock(joinPoint: ProceedingJoinPoint): Any {
        val signature = joinPoint.signature as MethodSignature
        val method = signature.method
        val redissonLock = method.getAnnotation<RedissonLock>(RedissonLock::class.java)

        val key = getDynamicValue(
            signature.parameterNames, joinPoint.args, redissonLock.key
        ) as String
        val rLock = redissonClient.getLock(key)

        try {
            val available = rLock.tryLock(
                redissonLock.waitTime, redissonLock.leaseTime, redissonLock.timeUnit
            )
            if (!available) {
                log.info("Lock 획득 실패 : {}", key)
                return false
            }

            return aopForTransaction.proceed(joinPoint, key)
        } catch (e: InterruptedException) {
            throw InterruptedException()
        } finally {
            try {
                log.info("Lock 해제 : {}", key)
                rLock.unlock()
            } catch (e: IllegalMonitorStateException) {
                log.info(
                    "이미 해제된 Lock : {} {}", method.name, key
                )
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(RedissonLockAop::class.java)
    }
}
